models/image.py

```diff
@@ -1242,6 +1077,244 @@ def load(self):
                 if not ( gotim and gotweight and gotflags ):
                     raise FileNotFoundError( "Failed to load at least one of image, weight, flags" )
 
+    def get_upstream_provenances(self):
+        """Collect the provenances for all upstream objects.
+
+        This will fail if any upstream objects with the same type have differing provenance IDs.
+        Note that upstream_images must each have the other related products
+        like sources, psf, wcs, etc. already loaded.
+        This happens when the objects are used to produce, e.g., a coadd or
+        a subtraction image, but they would not necessarily be loaded automatically from the DB.
+        To load those products (assuming all were previously committed with their own provenances)
+        use the load_upstream_products() method.
+
+        Returns
+        -------
+        list of Provenance objects:
+            a list of unique provenances, one for each data type.
+        """
+        image_provs = []
+        sources_provs = []
+        psf_provs = []
+        wcs_provs = []
+        zp_provs = []
+
+        for im in self.upstream_images:
+            image_provs.append(im.provenance)
+            if im.sources is not None:
+                sources_provs.append(im.sources.provenance)
+            if im.psf is not None:
+                psf_provs.append(im.psf.provenance)
+            if im.wcs is not None:
+                wcs_provs.append(im.wcs.provenance)
+            if im.zp is not None:
+                zp_provs.append(im.zp.provenance)
+
+        if len(set([p.id for p in image_provs])) > 1:
+            raise ValueError("Cannot combine images with different provenance IDs.")
+        if len(set([p.id for p in sources_provs])) > 1:
+            raise ValueError("Cannot combine sources with different provenance IDs.")
+        if len(set([p.id for p in psf_provs])) > 1:
+            raise ValueError("Cannot combine PSFs with different provenance IDs.")
+        if len(set([p.id for p in wcs_provs])) > 1:
+            raise ValueError("Cannot combine WCSs with different provenance IDs.")
+        if len(set([p.id for p in zp_provs])) > 1:
+            raise ValueError("Cannot combine ZPs with different provenance IDs.")
+
+        return image_provs + sources_provs + psf_provs + wcs_provs + zp_provs
+
+    def load_upstream_products(self, session=None):
+        """Make sure each upstream image has its related products loaded.
```

Comment:
Is the use case for this e.g. when you're making a coadd or a subtraction, and you want to have all the stuff for all the things that go into it?  I can see using it for that, but I'm not sure what else I'd use it for.